File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
def nPr(s,m):
    if s == m:
        result.append(li[0:m])
    else:
        for i in range(s,n):
            li[s],li[i] =li[i],li[s]
            nPr(s+1,m)
            li[s], li[i] = li[i], li[s]

n,m = map(int,input().split())
li = [i for i in range(1,n+1)]
result = []
nPr(0,m)

result.sort()
for i in result:
    print(*i)

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.*;
import java.util.*;
import java.util.stream.*;
public class Main {
    static Deque<Integer> result;
    static boolean[] visit;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] NM = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int N = NM[0];
        int M = NM[1];
        result = new ArrayDeque<>();
        visit = new boolean[N+1];
        int[] num_li = IntStream.rangeClosed(1,N).toArray();
        back(M,num_li);

    }
    public static void back(int M, int[] num_li) {
        if (result.size() == M) {
            System.out.println(result.stream().map(String::valueOf).collect(Collectors.joining(" ")));
            return ;
        }
        for (int num: num_li) {
            if (!visit[num]) {
                result.add(num);
                visit[num] = true;
                back(M,num_li);
                result.pollLast();
                visit[num] = false;
            }
        }
    }
}
